### PR TITLE
refactor(gemma4): the tests call the production path instead of shaping it

### LINKED ISSUE #945: Keep Gemma 4 ignored gates compilable and define their maintainer runner
## Problem

Gemma 4's checkpoint-backed gates are valuable but have no continuous compile contract or checked-in execution manifest.

At `69c4e149`:

- `pegainfer-gemma4` has 19 ignored library gates: 8 in `src/engine.rs` and 11 in `src/serve_oracle.rs`.
- The gates cover HF numeric parity, long/window context, generation, mixed admission, async prefill, prefix cache, CUDA Graph replay, cancellation, and lifecycle behavior.
- GitHub CI does not build or Clippy-check `pegainfer-gemma4 --features gemma4 --all-targets`.
- No checked-in runner names the maintained gate set.
- Engine gates document `--test-threads=1`, but a generic `--ignored` command does not encode the suite membership or prove that every expected gate still exists.

The tests currently fail closed when `PEGAINFER_TEST_MODEL_PATH` is absent and pin their checkpoint/fixtures. The decay risk is that feature-only test code can stop compiling or individual gates can fall out of the informal command set.

Related model roadmap: #758.

## Direction

Add feature-on compile/Clippy coverage without requiring a GPU in GitHub CI. Add one maintainer-run, fail-closed, serial command that owns the real checkpoint/GPU suite.

Do not split `serve_oracle.rs` in this issue. Establish the execution contract before reorganizing the files.

## Acceptance criteria

- GitHub CI runs release `cargo check` and Clippy with `-D warnings` for `pegainfer-gemma4 --features gemma4 --all-targets` using the repository's supported CUDA build setup.
- CI does not silently drop the feature when test targets are compiled.
- A checked-in maintainer runner explicitly names the 19 expected ignored gates, grouped by their invariant.
- The runner always uses `--test-threads=1` and fails before execution when the checkpoint, required fixture, pinned metadata, or CUDA device is unavailable.
- The runner prints the source commit, selected gate names, and selected/completed/failed counts; any missing or skipped gate makes the command fail.
- The runner serializes model ownership so two checkpoint-heavy tests are not launched concurrently.
- Existing external golden, row-ownership, lifecycle, prefix-cache, async-prefill, and graph/eager-equivalence coverage is retained.
- The PR includes one maintainer-run result for the exact head.

A GPU GitHub Actions runner is out of scope. This issue makes the manual GPU evidence reproducible and keeps the feature/test surface alive between runs.

## Contributor guardrails

- The PR description must map all 19 current gate names to an invariant group and show that the runner selects each name exactly once.
- Do not satisfy the runner contract with source grep, test-name wildcards, log-only `SKIP` detection, or an unbounded repository-wide `--ignored` command.
- Do not replace real checkpoint/GPU evidence with mocks, fake fixtures, simulated CUDA, or tests of the runner implementation itself.
- CI evidence must show feature-on all-target check and Clippy commands/results. A default-feature build is not equivalent.
- Keep the engine/oracle test bodies and production code unchanged except where required to make prerequisite failure explicit. File splitting, renaming, helper-framework extraction, and unrelated cleanup are out of scope.
- The exact-head maintainer transcript must include selected/completed/failed counts and the serial setting; a PR-body claim without command output is insufficient.

### PR DESCRIPTION
**Pull request by FeathBow**
## Description

Closes #945

This removes the production code that existed for gates, merges gates that proved the same thing twice, and gives the environment boundary one failure semantic.

**Production stops carrying test shapes.** `LogitsSpan` is gone: its `All` variant was constructed in one ignored test while serving only ever read the last row, so `step` now materializes that row unconditionally and the oracle compares exactly what production computes. The window-release flag keeps its field but loses its five `#[cfg(test)]` / `#[cfg(not(test))]` branch pairs in favour of one `advance_local` helper, so the test build and the release build compile the same frontier-advance logic. `full_forward`, the layer probes and their imports move behind `#[cfg(test)]`, which lets the crate drop all five `#[allow(dead_code)]` allowances.

**Suppression moves to a kernel because a duplicate check is gone.** The host path re-validated every suppressed id against the head width on every call, duplicating the check `GenerationPolicy::check_against_vocab` already performs at load. With that duplicate removed the ids can live on the device pre-validated, so `rows × ids` device-to-device copies collapse into one launch of a new shared `suppress_logits_bf16_in_place`, whose Rust wrapper documents the caller's obligation and whose launcher refuses non-positive geometry and row-count overflow.

**Gates merge, and arithmetic leaves the device.** The two long-context waypoint gates share one tower load and one table-driven body; the three engine lifecycle scripts become one test that runs the shared, green and gathered variants in sequence; the raised-ceiling assertion folds into the roster-edge gate. The segmented-admission ledger gate no longer loads a 12B checkpoint to prove page bookkeeping — `plan_release` is now a pure function over a small state, and two CPU unit tests hold whole-prompt and segmented admission to the identical ledger with a residency bound. `resume_point` receives the same treatment: the prefix-cache test used to re-implement the resume floor with a different constant than production, and now calls it.

**The knob boundary gets one shape.** All five readers go through one `normalize_env`, which is the single place that maps a missing variable to a default and a non-UTF-8 one to a refusal. `PEGAINFER_PREFIX_CACHE` joins the others in refusing malformed values instead of silently disabling itself, while keeping the spelling of off the chunked walk already used — unset, empty, `0` and `off` all mean "allocate nothing", so no existing configuration changes behaviour. `AsyncPrefillMode`'s `Off` variant becomes `Option<LaneMode>`, which deletes the unrepresentable-state guard it needed.

**CI covers the model line.** The Qwen3.5 CUDA job generalizes into a two-model matrix, so Gemma 4 gets `check` and `clippy -D warnings` on sm_80 alongside it, with the Triton install now conditional on the model that needs it.

### Test Env

- Single GPU (sm_89, x86_64), CUDA 12.9, Gemma 4 12B checkpoint.

### Verification

- Formatting and the kernels crate's own targets are clean: `cargo fmt --all --check`, and `cargo check -p pegainfer-kernels --all-targets`, which is what compiles the device-gate test binaries this change edits.
- The crate's CPU tests pass: 40 passed, 0 failed, 17 ignored — including the new knob-boundary tests, the release-plan and resume-point unit tests that replaced two device gates, and the pool arithmetic.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -320,22 +320,41 @@ jobs:
           PEGAINFER_CUDA_SM: "80"
           PEGAINFER_NVCC_JOBS: "2"
 
-  qwen35-cuda:
+  feature-model-cuda:
```
[the remark this answers is not in the document]
Comment (FeathBow) [reply .github/workflows/ci.yml:323]: DCO passed.

**Comment by FeathBow** [comment]
The preflight is tiered. Each manifest entry declares what its gate needs — gpu, ckpt, prompts, fixtures, chatgolden — and a run demands only the union over what the filter selected. Shown rather than asserted:

$ env -u PEGAINFER_TEST_MODEL_PATH scripts/gemma4_gates.sh suppression
gemma4 gates: prerequisites gpu
gemma4 gates: 1 selected of 20 in the manifest
gemma4 gates: selected 1, completed 1, failed 0

$ scripts/gemma4_gates.sh string_form
gemma4 gates: prerequisites ckpt chatgolden
     Running tests/tokenizer_parity.rs
gemma4 gates: selected 1, completed 1, failed 0

The first runs with no checkpoint in the environment at all; before this it died on the opening preflight statement. Two entries are not what their group suggests, and both were read out of the code rather than assumed: the roster-edge gate reaches the generate fixture through walk_prompts, and the_raise_refuses_without_its_prerequisites needs only the config, because EngineState::load settles the knob refusals above both the weight load and the device open. The preflight also moved after selection, so manifest drift now fails before anything demands a 22 GiB checkpoint.
The multi-target runner has a transcript. At 7d33159f, the commit with nothing modified, single GPU (sm_89,x86_64): cargo check and cargo clippy -- -D warnings both clean, and 

preflight: 4 fixtures agree on revision 707f0a3b
gemma4 gates: prerequisites gpu ckpt prompts fixtures chatgolden
gemma4 gates: source 7d33159f3286a8dabfa9abe3283c06eb12beba2d                                                                     gemma4 gates: 20 selected of 20 in the manifest
gemma4 gates: selected 20, completed 20, failed 0 

on a card checked quiet beforehand, 186 MiB in use at bbinary is inside that count.

The KV ledger tests are deleted rather than rewritten. mean giving production a seam for the test to share,which is the shaping this branch exists to undo — and it is unnecessary: gate_waypoint already pins the resident page count exactlagainst a closed form derived from kv_len alone, for boe chunked form of the same waypoint, throughadvance_and_release itself. Whole and segmented release cannot part ways without one of those failing. What went with them is the single-walker residency peak bound, which they alone co the mirror rather than the real application. Neithertest was executed by anything in any case: both are platest matrix does not carry this crate, and its CUDA jobsonly compile.
